### DISCUSSION TITLE
Collect, document, and answer FAQ

### DISCUSSION THREAD

**Discussion by SergioBenitez** [rwf2/Rocket#1836] [General]
There are questions that are frequently asked that are not directly related to Rocket but impact Rocketeers nevertheless. In pursuit of making Rocket as friendly and easy to use, we should collect these questions in an official capacity and answer them once and for all. This issue tracks such questions.

  * [ ] How do I listen on IPv4 and IPv6?
  * [x] How do I upload a file / what is this multipart metadata in my stream?
  * [x] Can I, and if so how, do I handle websockets?
  * [x] How do I respond with different kinds of responses, with different status codes?
  * [x] Should I use global state with something like `lazy_static!`?
  * [x] How do I access managed state outside of a Rocket-related context?
  * [ ] How do I mount a generic route?
  * [x] How do I make Rocket restart automatically when I change source code?
  * [x] How do I get an `&Request` in a handler?
  * [ ] How do I get a specific header in a handler?
  * [x] When will version `x + 1` be released? Why does it take so long?
  * [x] Is example `foo` broken? It doesn't work for me.
  * [ ] Why so many macros? I heard they lead to bad errors and make my code harder to understand.
  * [x] Is Rocket slow? Is Rocket fast?
  * [x] Can I use Rocket in production? Should I? It's only 0.x!
 
Please feel free to make suggestions!

**Comment by ThouCheese** [comment]
I would suggest adding "why doesn't writing `rocket = "0.5.0-dev"` in Cargo.toml work"

**Comment by jebrosen** [comment]
* How do I get the value of a request header (e.g. `Host`, `User-Agent`) in a route?

**Comment by Flying-Toast** [comment]
- Why is {x} in Rocket slower than {other framework}

**Comment by SergioBenitez** [comment]
Hiding comments as "resolved" as I bring them into the top-level post or they get addressed elsewhere in docs.

**Comment by Darthfett** [comment]
- How do I configure Rocket to handle SSL traffic (HTTPS)?

**Comment by Xunjin** [comment]
How can I structure a more "long term" project? Focusing on maintenance and segment concerns (maybe some example in "known" software architectures). 

PS: That would probably be a book/set of articles itself.

**Reply by SergioBenitez** [reply]
Indeed, this is a question I'd love to answer but one that requires more than a few paragraphs. I (or someone else) will eventually write such a book, I imagine.

**Comment by jhwgh1968** [comment]
* How do I start and manage rocket from non-async Rust code?

This is the question I had while upgrading a rather complex project from 0.4.

In this project, a lot of other stuff was going on, and the HTTP server was only a secondary function enabled when requested. Converting the rest of my code to async Rust was simply too much work.

Instead, I wrote a `start_server` function to replicate the conditional config and blocking startup in pre-async Rocket. But I feel pretty certain that ugly hack is the wrong approach.

Not only does it not use the new `#[launch]` macro anywhere, but I had to use `unsafe` code to actually return errors properly. (I wanted to get the `std::io::Error` from a `rocket::Error` of `Kind::Io` by value, but the `ErrorKind` that contains it is returned by reference.)

I imagine a Migration Guide will need to be written at some point, and I would offer to provide more details elsewhere if it would help with that effort. This is just to register that I think a question providing a hint/official pointer should be in the FAQ.

**Reply by SergioBenitez** [reply]
The answer [is here](https://rocket.rs/master/guide/faq/#internal-server).

> Not only does it not use the new `#[launch]` macro anywhere, but I had to use `unsafe` code to actually return errors properly. (I wanted to get the `std::io::Error` from a `rocket::Error` of `Kind::Io` by value, but the `ErrorKind` that contains it is returned by reference.)

This is totally unnecessary. `Error` and `io::Error` implement `std::error::Error`, so at minimum, you can return a `Box<std::error::Error>`. Better still, you can create your own `enum` with the possible variants.

**Reply by jhwgh1968** [reply]
> Better still, you can create your own enum with the possible variants.

Only today did I realize that I should keep the `rocket::Error` object inside my own enum, and take it apart in a custom `source` method.

It's a bit awkward having an extra case of `RocketError` instead of being able to fold it into the same field as `std::io::Error`, but I suppose that lets me treat the enum items as indicating where I am in the process. It's a better design after all!

Thanks for the pointers!

EDIT: better design

**Reply by the10thWiz** [reply]
@jhwgh1968 I would recommend using a more regular approach to starting the server: namely, by using `#[launch]` or `#[rocket::main]`, just from another module, and calling the generated main function in a new thread to start the server. You can then use the existing shutdown mechanisms to shutdown the server when it's done.

Using the codegen (`launch` or `rocket::main`), is highly recommended, because it makes maintaining the code simpler; the internal functions to start the server are considered somewhat unsafe, and subject to change. Relying on the codegen prevents changes to these internal APIs affecting your codebase, since the codegen will simply output the correct code to start the server.

**Reply by MikailBag** [reply]
Here is a way to run Rocket server on an already existing Tokio runtime
```rust
async fn run_rocket_server() -> Result<(), Box<dyn std::error::Error>> {
    let rocket = rocket::build();
    // configure routes, fairing, etc
    rocket.launch().await?;
    Ok(())
}
```
This code does not use any internal APIs.

**Reply by SergioBenitez** [reply]
> Here is a way to run Rocket server on an already existing Tokio runtime
> 
> ```rust
> 
> async fn run_rocket_server() -> Result<(), Box<dyn std::error::Error>> {
> 
>     let rocket = rocket::build();
> 
>     // configure routes, fairing, etc
> 
>     rocket.launch().await?;
> 
>     Ok(())
> 
> }
> 
> ```
> 
> This code does not use any internal APIs.

This is fine, but it isn't recommended. The Rocket-created runtime is tuned for Rocket applications. It's also Rocket-aware with regards to shutdown. If you do this, graceful shutdown may not work, as indicated in the documentation.

**Reply by the10thWiz** [reply]
I believe running Rocket on an existing runtime is explicitly not recommended.

From my understanding, Rocket hopes to eventually be able to switch to a different runtime than Tokio, and code like this would make that a breaking change.

**Reply by SergioBenitez** [reply]
See also https://api.rocket.rs/v0.5-rc/rocket/fn.execute.html.

**Comment by pzingg** [comment]
How do I pre-populate a Form Context with values? The `forms` example (which uses a `values_for` Tera macro) only shows how to set an empty Context (with `Context::default()`). What if I want to have some actual non-empty, non-missing values on the first render (before the Form is submitted)? I don’t see how one could set values in the Form’s inner data and convert them into a Context, with the same `values` and `errors` structure as when the Form data is parsed in a POST handler.

**Reply by SergioBenitez** [reply]
`Context` is only designed to be used as the error type of a failed form parse. This is why, for instance, it doesn't implement `FromForm`. Instead, Rocket enforces strictly typed form parsing as indicated throughout the docs.

Thus, pre-population of a form should be done by populating a strongly typed value with whatever field values you'd like and rendering those fields in the template.
